postgres_ffi/src/lib.rs

```diff
@@ -14,3 +14,43 @@ pub mod nonrelfile_utils;
 pub mod pg_constants;
 pub mod relfile_utils;
 pub mod xlog_utils;
+
+//  See TransactionIdIsNormal in transam.h
+pub const fn transaction_id_is_normal(id: TransactionId) -> bool {
+    id > pg_constants::FIRST_NORMAL_TRANSACTION_ID
+}
+
+// See TransactionIdPrecedes in transam.c
+pub const fn transaction_id_precedes(id1: TransactionId, id2: TransactionId) -> bool {
+    /*
+     * If either ID is a permanent XID then we can just do unsigned
+     * comparison.  If both are normal, do a modulo-2^32 comparison.
+     */
+
+    if !(transaction_id_is_normal(id1)) || !transaction_id_is_normal(id2) {
+        return id1 < id2;
+    }
+
+    let diff = id1.wrapping_sub(id2) as i32;
+    return diff < 0;
+}
+
+// See CLOGPagePrecedes in clog.c
+pub const fn clogpage_precedes(page1: u32, page2: u32) -> bool {
+    let mut xid1 = page1 * pg_constants::CLOG_XACTS_PER_PAGE;
+    xid1 += pg_constants::FIRST_NORMAL_TRANSACTION_ID + 1;
+    let mut xid2 = page2 * pg_constants::CLOG_XACTS_PER_PAGE;
+    xid2 += pg_constants::FIRST_NORMAL_TRANSACTION_ID + 1;
+
+    transaction_id_precedes(xid1, xid2)
+        && transaction_id_precedes(xid1, xid2 + pg_constants::CLOG_XACTS_PER_PAGE - 1)
+}
+
+// See SlruMayDeleteSegment() in slru.c
+pub fn slru_may_delete_clogsegment(segpage: u32, cutoff_page: u32) -> bool {
```

Comment:
not const because assertions cannot be const yet, but there is a crate [static_assertions](https://docs.rs/static_assertions/1.1.0/static_assertions/macro.const_assert.html) is it worth using it? If this is the only place probably not

Reply:
static_assertions won't help here anyway, because `segpage` is not a constant.

```rust
// cargo-deps: static_assertions = "*"
use static_assertions::const_assert;

const fn test(x: i32) {
    const_assert!(x == 0);
}
```

```
error[E0435]: attempt to use a non-constant value in a constant
 --> test.rs:5:19
  |
4 | const fn test(x: i32) {
  |               - this would need to be a `const`
5 |     const_assert!(x == 0);
  |                   ^

error: aborting due to previous error
```

Reply:
Meh, indeed

Reply:
This is caused by the fact that const functions are not "const-only" but rather "also-const". The function may not know the value of a runtime parameter.

Reply:
That's reasonable